### ISSUE TITLE
Provide public API equivalent for DefaultSourceDirectorySet

### ISSUE THREAD

**Issue by eriwen** [gradle/gradle#727]
This is essential for any sort of language support. Without this being public (or rather some sort of public factory), it's almost impossible to make a new language SourceSet. Note that the new model has a factory for this, but there is not one for the current model.

**Comment by jjohannes** [comment]
This is also important for the use case where you want to build a proper Java module that is still compiled with Java8, such that it can be used as a normal library by Java8 consumers.

Then you want this setup:

```
src
└── main
    ├── java
    │   └── (all your sources)
    └── java9
        └── module-info.java
```

**Comment by melix** [comment]
@jjohannes Isn't it already available via `ObjectFactory#sourceDirectorySet`? Or am I missing something?

**Comment by jjohannes** [comment]
Yes. We have this API to create an instance. But how do you wire that in to solve the actual use case(s). This is the simplest I could manage:

https://github.com/jjohannes/gradle-jpms-experiments/blob/afd93d810f8f1a077a660c4791a5bce24b798bf1/module-java8/build.gradle.kts#L12-L39

But it uses internal API and is very verbose.

**Comment by melix** [comment]
Right, but I think it's orthogonal to the original ticket, isn't it? Should we rename?

**Comment by jjohannes** [comment]
Feel free to rename this if you have a good name :)

I think this sould probably be addressed by providing some form of API to do the wiring here:
```
var sourceSetOutput = main.output as org.gradle.api.internal.tasks.DefaultSourceSetOutput
sourceSetOutput.addClassesDir { java9.destinationDirectory.asFile.get() }
sourceSetOutput.registerCompileTask(compileJava9)
java9.compiledBy(compileJava9, AbstractCompile::getDestinationDirectory)
```

It should not be necessary to have the  `registerCompileTask` and `addClassesDir` which work with tasks/non-lazy API. It should be enough to wire the _properties_.

And there should be some way to create the thing and directly do this wiring to the source set it belongs to:
```
val java9 = objects.sourceDirectorySet("java9", "Java 9 Sources")
java9.srcDir("src/" + main.name + "/" + java9.name)
main.allJava.source(java9)
main.allSource.source(java9)
val sourceSetChildPath = "classes/" + java9.name + "/" + main.name
```
Maybe the source set could offer a factory method for that.

It might be good to spike this maybe write a short spec.

I think if the API we come up with lets you conveniently address the "Java 9" use case, it would be a good start. Ideally it should be useable by other Plugins like Kotlin and Android to do this kind of thing with the public API. And our own plugins (groovy, scala) should then also use that instead of the internal APIs they use now.

**Comment by melix** [comment]
Spiked an API to register a language directory source set in https://github.com/gradle/gradle/pull/13395/commits/56bcb0d17f2ef6bc1132fa211edd201cd8ecf215, with a real word example [here](https://github.com/melix/rigel/blob/cc/high-level-java-api/buildSrc/src/main/kotlin/me.champeau.module.gradle.kts#L5-L15)

**Comment by melix** [comment]
I think this is still relevant, should be reopened.

**Comment by mattbertolini** [comment]
I have recently encountered this issue while writing a custom Gradle plugin. I'm using a custom `SourceDirectorySet` and would like to extend the DSL to add custom sections for source set specific configuration, rather than duplicating the logic in another area or having them configure the specific task (which requires them knowing the task name). In order to do that I have to extend the `SourceDirectorySet` and by extension the `DefaultSourceDirectorySet`:

```java
public class DefaultCustomPluginSourceDirectorySet extends DefaultSourceDirectorySet implements CustomPluginSourceDirectorySet {
    private final GeneratorContainer generatorContainer;

    @Inject
    public DefaultCustomPluginSourceDirectorySet(SourceDirectorySet sourceSet, TaskDependencyFactory taskDependencyFactory, ObjectFactory objectFactory) {
        super(sourceSet, taskDependencyFactory);
        generatorContainer = objectFactory.newInstance(GeneratorContainer.class);
    }

    // Extending the DSL here
    @Override
    public GeneratorContainer getGenerators() {
        return generatorContainer;
    }

    @Override
    public void generators(final Action<GeneratorContainer> action) {
        action.execute(generatorContainer);
    }
}
```

I don't like importing internal packages but in this case I have no choice. This would be solved by having a public API equivalent of the `DefaultSourceDirectorySet`. I guess it could also be solved by making the `SourceDirectorySet` implement the `ExtensionAware` interface so I can just extend the implementation generated by the `ObjectFactory`. Just my two cents.
